component/sqs/component.go

```diff
@@ -0,0 +1,332 @@
+	nackMessageState    messageState = "NACK"
+	fetchedMessageState messageState = "FETCHED"
+)
+
+var (
+	messageAge     *prometheus.GaugeVec
+	messageCounter *prometheus.CounterVec
+	queueSize      *prometheus.GaugeVec
+)
+
+func init() {
+	messageAge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_age",
+			Help:      "Message age based on the SentTimestamp SQS attribute",
+		},
+		[]string{"sqsAPI"},
+	)
+	prometheus.MustRegister(messageAge)
+	messageCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_counter",
+			Help:      "Message counter",
+		},
+		[]string{"sqsAPI", "state", "hasError"},
+	)
+	prometheus.MustRegister(messageCounter)
+	queueSize = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "queue_size",
+			Help:      "Queue size reported by AWS",
+		},
+		[]string{"state"},
+	)
+	prometheus.MustRegister(queueSize)
+}
+
+// Component implementation of a async component.
+type Component struct {
+	name              string
+	queueName         string
+	queueURL          string
+	sqsAPI            sqsiface.SQSAPI
+	maxMessages       *int64
+	pollWaitSeconds   *int64
+	visibilityTimeout *int64
+	statsInterval     time.Duration
+	proc              ProcessorFunc
+	retries           uint
+	retryWait         time.Duration
+}
+
+// New creates a new component with support for functional configuration.
+func New(name, queueName string, sqsAPI sqsiface.SQSAPI, proc ProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	if name == "" {
+		return nil, errors.New("component name is empty")
+	}
+
+	if queueName == "" {
+		return nil, errors.New("queue name is empty")
+	}
+
+	if sqsAPI == nil {
+		return nil, errors.New("SQS API is nil")
+	}
+
+	if proc == nil {
+		return nil, errors.New("process function is nil")
+	}
+
+	out, err := sqsAPI.GetQueueUrlWithContext(context.Background(), &sqs.GetQueueUrlInput{
+		QueueName: aws.String(queueName),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get queue URL: %w", err)
+	}
+
+	cmp := &Component{
+		name:              name,
+		queueName:         queueName,
+		queueURL:          aws.StringValue(out.QueueUrl),
+		sqsAPI:            sqsAPI,
+		maxMessages:       aws.Int64(3),
+		pollWaitSeconds:   nil,
+		visibilityTimeout: nil,
+		statsInterval:     10 * time.Second,
+		proc:              proc,
+		retries:           10,
+		retryWait:         time.Second,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the consumer processing loop messages.
+func (c *Component) Run(ctx context.Context) error {
+	chErr := make(chan error)
+
+	go c.consume(ctx, chErr)
+
+	tickerStats := time.NewTicker(c.statsInterval)
+	defer tickerStats.Stop()
+	for {
+		select {
```

Comment:
it could be an option to log the case here for Done and err ...
This would happen only once i assume, correct ?

Reply:
err is logged at patron service level. Done could be interresting though.

Reply:
done